packages/astro/src/assets/fonts/vite-plugin-fonts.ts

```diff
@@ -0,0 +1,237 @@
+import { removeTrailingForwardSlash } from '@astrojs/internal-helpers/path';
+import type { Logger } from '../../core/logger/core.js';
+import { createCache, createStorage, type Cache } from './cache.js';
+
+interface Options {
+	settings: AstroSettings;
+	sync: boolean;
+	logger: Logger;
+}
+
+type PreloadData = Array<{ url: string; type: string }>;
+
+const createLogManager = (logger: Logger) => {
+	const done = new Set<string>();
+	const items = new Set<string>();
+	let id: NodeJS.Timeout | null = null;
+
+	return {
+		add: (value: string) => {
+			if (done.has(value)) {
+				return;
+			}
+
+			if (items.size === 0 && id === null) {
+				logger.info('assets', 'Downloading fonts...');
+			}
+			items.add(value);
+			if (id) {
+				clearTimeout(id);
+				id = null;
+			}
+		},
+		remove: (value: string, cached: boolean) => {
+			if (done.has(value)) {
+				return;
+			}
+
+			items.delete(value);
+			done.add(value);
+			if (id) {
+				clearTimeout(id);
+				id = null;
+			}
+			id = setTimeout(() => {
+				let msg = 'Done';
+				if (cached) {
+					msg += ' (loaded from cache)';
+				}
+				logger.info('assets', msg);
+			}, 50);
+		},
+	};
+};
+
+export function fonts({ settings, sync, logger }: Options): Plugin {
+	if (!settings.config.experimental.fonts) {
+		// this is required because the virtual module does not exist
+		// when fonts are not enabled, and that prevents rollup from building
+		// TODO: remove once fonts are stabilized
+		return {
+			name: 'astro:fonts:fallback',
+			config() {
+				return {
+					build: {
+						rollupOptions: {
+							external: [VIRTUAL_MODULE_ID],
+						},
+					},
+				};
+			},
+		};
+	}
+
+	const providers: Array<FontProvider<string>> = settings.config.experimental.fonts.providers ?? [];
+	const families: Array<FontFamily<string>> = settings.config.experimental.fonts.families;
+
+	const baseUrl = removeTrailingForwardSlash(settings.config.base) + URL_PREFIX;
+
+	let resolvedMap: Map<string, { preloadData: PreloadData; css: string }> | null = null;
+	/** Key is `${hash}.${ext}`, value is a URL */
+	let collected: Map<string, string> | null = null;
+	let isBuild: boolean;
+	let cache: Cache['cache'] | null = null;
+
+	return {
+		name: 'astro:fonts',
+		config(_, { command }) {
+			isBuild = command === 'build';
+		},
+		async buildStart() {
+			const { h64ToString } = await xxhash();
+
+			const resolved = await resolveProviders({
+				settings,
+				providers,
+			});
+
+			const storage = createStorage({
+				// In build, first check the build cache then fallback to dev cache
+				// In dev, only check dev cache
+				bases: [
+					new URL(CACHE_DIR, settings.dotAstroDir),
+					...(isBuild ? [new URL(CACHE_DIR, settings.config.cacheDir)] : []),
+				],
+			});
+			cache = createCache({ storage }).cache;
+
+			const { resolveFont } = await unifont.createUnifont(
+				resolved.map((e) => e.provider(e.config)),
+				{ storage },
+			);
+			resolvedMap = new Map();
+			collected = new Map();
```

Comment:
If we assign an empty map here, wouldn't make sense to assign the empty map when we define the variable? Unless `undefined` and empty map have different semantics here, but since the code never checks of a map being empty, we can safely initialise the empty map